### PR TITLE
Deny logging artifacts if user does not have permission to update run

### PR DESCRIPTION
**Pull request by harupy**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

Deny logging artifacts if user does not have permission to update run.

## How is this patch tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by harupy** [comment]
Interesting. `Run with id=e8f3b6e8681e4ef3be4c094586aa2fff` should exist, but not found.

```
    def test_log_artifact(client, monkeypatch, tmp_path):
        username1, password1 = create_user(client.tracking_uri)
        username2, password2 = create_user(client.tracking_uri)
        with User(username1, password1, monkeypatch):
            exp_id = client.create_experiment("exp")
            run = client.create_run(experiment_id=exp_id)
    
        tmp_file = tmp_path / "test.txt"
        tmp_file.touch()
        with User(username2, password2, monkeypatch):
            with pytest.raises(MlflowException, match=r"Permission denied"):
>               client.log_artifact(run.info.run_id, tmp_file)
E               AssertionError: Regex pattern did not match.
E                Regex: 'Permission denied'
E                Input: 'RESOURCE_DOES_NOT_EXIST: Run with id=e8f3b6e8681e4ef3be4c094586aa2fff not found'

client     = <mlflow.tracking.client.MlflowClient object at 0x7f665a2a6fa0>
exp_id     = '1'
monkeypatch = <tests.conftest.ExtendedMonkeyPatch object at 0x7f665a2a6070>
password1  = '860cf00a86'
password2  = '2e8e4e0a2a'
run        = <Run: data=<RunData: metrics={}, params={}, tags={'mlflow.runName': 'illustrious-toad-136'}>, info=<RunInfo: artifact_uri='file:///tmp/pytest-of-runner/pytest-0/test_log_artifact0/artifacts/1/e8f3b6e8681e4ef3be4c094586aa2fff/artifacts', end_time=None, experiment_id='1', lifecycle_stage='active', run_id='e8f3b6e8681e4ef3be4c094586aa2fff', run_name='illustrious-toad-136', run_uuid='e8f3b6e8681e4ef3be4c094586aa2fff', start_time=1687745451272, status='RUNNING', user_id='unknown'>, inputs=<RunInputs: dataset_inputs=[]>>
tmp_file   = Path('/tmp/pytest-of-runner/pytest-0/test_log_artifact0/test.txt')
tmp_path   = Path('/tmp/pytest-of-runner/pytest-0/test_log_artifact0')
username1  = '91986b854c'
username2  = '336d686e21'
```

**Comment by gabrielfu** [comment]
> Interesting. `Run with id=e8f3b6e8681e4ef3be4c094586aa2fff` should exist, but not found.
> 
> ```
>     def test_log_artifact(client, monkeypatch, tmp_path):
>         username1, password1 = create_user(client.tracking_uri)
>         username2, password2 = create_user(client.tracking_uri)
>         with User(username1, password1, monkeypatch):
>             exp_id = client.create_experiment("exp")
>             run = client.create_run(experiment_id=exp_id)
>     
>         tmp_file = tmp_path / "test.txt"
>         tmp_file.touch()
>         with User(username2, password2, monkeypatch):
>             with pytest.raises(MlflowException, match=r"Permission denied"):
> >               client.log_artifact(run.info.run_id, tmp_file)
> E               AssertionError: Regex pattern did not match.
> E                Regex: 'Permission denied'
> E                Input: 'RESOURCE_DOES_NOT_EXIST: Run with id=e8f3b6e8681e4ef3be4c094586aa2fff not found'
> 
> client     = <mlflow.tracking.client.MlflowClient object at 0x7f665a2a6fa0>
> exp_id     = '1'
> monkeypatch = <tests.conftest.ExtendedMonkeyPatch object at 0x7f665a2a6070>
> password1  = '860cf00a86'
> password2  = '2e8e4e0a2a'
> run        = <Run: data=<RunData: metrics={}, params={}, tags={'mlflow.runName': 'illustrious-toad-136'}>, info=<RunInfo: artifact_uri='file:///tmp/pytest-of-runner/pytest-0/test_log_artifact0/artifacts/1/e8f3b6e8681e4ef3be4c094586aa2fff/artifacts', end_time=None, experiment_id='1', lifecycle_stage='active', run_id='e8f3b6e8681e4ef3be4c094586aa2fff', run_name='illustrious-toad-136', run_uuid='e8f3b6e8681e4ef3be4c094586aa2fff', start_time=1687745451272, status='RUNNING', user_id='unknown'>, inputs=<RunInputs: dataset_inputs=[]>>
> tmp_file   = Path('/tmp/pytest-of-runner/pytest-0/test_log_artifact0/test.txt')
> tmp_path   = Path('/tmp/pytest-of-runner/pytest-0/test_log_artifact0')
> username1  = '91986b854c'
> username2  = '336d686e21'
> ```
> 
> https://github.com/mlflow/mlflow/actions/runs/5372974379/jobs/9746897484?pr=8817
> 
> @gabrielfu Any clue on how this happened?

that's interesting, I'm debugging locally now

File: mlflow/tracking/_tracking_service/client.py [hunk 1]
```diff
@@ -438,13 +438,22 @@ def _get_artifact_repo(self, run_id):
             TrackingServiceClient._artifact_repos_cache[run_id] = artifact_repo
             return artifact_repo
 
+    def _validate_can_update_run(self, run_id):
+        self.store.update_run_info(
+            run_id=run_id,
+            run_status=None,
+            end_time=None,
+            run_name=None,
+        )
```
Comment (harupy) [inline mlflow/tracking/_tracking_service/client.py:447]: TODO: Insert a comment on why we need this.

**Comment by gabrielfu** [comment]
@harupy I located the bug at 
https://github.com/mlflow/mlflow/blob/8773d89cb467ec1ba44771e137617ccd29b11221/mlflow/server/auth/__init__.py#L215

where it's fetching from REST store again

I will find a solution and file a PR

**Comment by gabrielfu** [comment]
@harupy found that auth test setup is using a different db from the main mlflow db, fixed at #8837

**Comment by harupy** [comment]
#8837 is merged, rebased on master, the failed test should pass this time.

**Pull request closed without merging**